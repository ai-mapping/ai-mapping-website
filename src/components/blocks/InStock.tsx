import Link from 'next/link';
import React from 'react';
import InStockTable from './InStockTable';
import Section from '@/components/layout/Section';

// Define types for table columns
export type TableColumn = {
  key: string;
  header: string;
  align?: 'left' | 'center' | 'right';
};

// Define types for stock items
export type StockItem = {
  id: string;
  name: string;
  partNumber: string;
  compatibility: string;
  inStock: 'available' | 'limited' | 'unavailable';
};

// Main props interface
export type InStockProps = {
  heading: string;
  subheading: string;
  columns: TableColumn[];
  items: StockItem[];
  orderButtonText: string;
  viewAllLinkText: string;
  viewAllLink: string;
};

// Default props
export const defaultInStockProps: InStockProps = {
  heading: 'Components Library',
  subheading: 'Ready-to-use components for your next project',
  columns: [
    { key: 'name', header: 'Component Name', align: 'left' },
    { key: 'partNumber', header: 'Version', align: 'left' },
    { key: 'compatibility', header: 'Framework Support', align: 'left' },
    { key: 'inStock', header: 'Status', align: 'center' },
    { key: 'action', header: 'Action', align: 'center' }
  ],
  items: [
    {
      id: '1',
      name: 'AI Form Builder',
      partNumber: 'v1.2.0',
      compatibility: 'React, Next.js, Vue',
      inStock: 'available'
    },
    {
      id: '2',
      name: 'Dynamic Hero Section',
      partNumber: 'v2.0.1',
      compatibility: 'React, Next.js, Astro',
      inStock: 'available'
    },
    {
      id: '3',
      name: 'Testimonial Carousel',
      partNumber: 'v1.5.2',
      compatibility: 'React, Vue, Svelte',
      inStock: 'available'
    },
    {
      id: '4',
      name: 'AI Content Generator',
      partNumber: 'v0.9.8',
      compatibility: 'React, Next.js',
      inStock: 'limited'
    }
  ],
  orderButtonText: 'Use Component',
  viewAllLinkText: 'View all available components',
  viewAllLink: '/components'
};

const InStock = ({
  heading = defaultInStockProps.heading,
  subheading = defaultInStockProps.subheading,
  columns = defaultInStockProps.columns,
  items = defaultInStockProps.items,
  orderButtonText = defaultInStockProps.orderButtonText,
  viewAllLinkText = defaultInStockProps.viewAllLinkText,
  viewAllLink = defaultInStockProps.viewAllLink
}: Partial<InStockProps>) => {
  return (
    <Section background="surface" paddingY="lg" containerWidth="container">
        <h2 className="text-2xl md:text-3xl font-semibold text-center mb-3 text-white">
          {heading}
        </h2>
        <p className="text-neutral text-center mb-10 max-w-2xl mx-auto">
          {subheading}
        </p>

        <div className="overflow-x-auto">
          <InStockTable 
            columns={columns} 
            items={items} 
            orderButtonText={orderButtonText} 
          />
        </div>

        <div className="text-center mt-8">
          <Link
            href={viewAllLink}
            className="text-primary font-medium hover:text-primary-light transition-colors duration-300"
          >
            {viewAllLinkText}
          </Link>
        </div>
    </Section>
  );
};

export default InStock;
