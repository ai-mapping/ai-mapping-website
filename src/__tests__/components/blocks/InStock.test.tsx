import React from 'react';
import { render, screen, within } from '@testing-library/react';
import InStock, { defaultInStockProps } from '@/components/blocks/InStock';

// Mock the InStockTable component
jest.mock('@/components/blocks/InStockTable', () => {
  return {
    __esModule: true,
    default: ({ columns, items, orderButtonText }) => (
      <div data-testid="in-stock-table">
        <div>Columns: {columns.length}</div>
        <div>Items: {items.length}</div>
        <div>Order Button Text: {orderButtonText}</div>
        <div>
          {items.map((item, index) => (
            <div key={index} data-testid={`stock-item-${index}`}>
              <div>Name: {item.name}</div>
              <div>Part Number: {item.partNumber}</div>
              <div>Compatibility: {item.compatibility}</div>
              <div>In Stock: {item.inStock}</div>
            </div>
          ))}
        </div>
      </div>
    )
  };
});

describe('InStock Component', () => {
  it('renders with default props correctly', () => {
    const { container } = render(<InStock />);
    
    // Check if heading and subheading are rendered
    expect(screen.getByText(defaultInStockProps.heading)).toBeInTheDocument();
    expect(screen.getByText(defaultInStockProps.subheading)).toBeInTheDocument();
    
    // Check if the InStockTable is rendered
    expect(screen.getByTestId('in-stock-table')).toBeInTheDocument();
    expect(screen.getByText(`Columns: ${defaultInStockProps.columns.length}`)).toBeInTheDocument();
    expect(screen.getByText(`Items: ${defaultInStockProps.items.length}`)).toBeInTheDocument();
    expect(screen.getByText(`Order Button Text: ${defaultInStockProps.orderButtonText}`)).toBeInTheDocument();
    
    // Check if all stock items are rendered
    defaultInStockProps.items.forEach((item, index) => {
      const stockItem = screen.getByTestId(`stock-item-${index}`);
      expect(stockItem).toBeInTheDocument();
      expect(within(stockItem).getByText(`Name: ${item.name}`)).toBeInTheDocument();
      expect(within(stockItem).getByText(`Part Number: ${item.partNumber}`)).toBeInTheDocument();
      expect(within(stockItem).getByText(`Compatibility: ${item.compatibility}`)).toBeInTheDocument();
      expect(within(stockItem).getByText(`In Stock: ${item.inStock}`)).toBeInTheDocument();
    });
    
    // Check if "view all" link is rendered
    expect(screen.getByText(defaultInStockProps.viewAllLinkText)).toBeInTheDocument();
    expect(screen.getByText(defaultInStockProps.viewAllLinkText).closest('a')).toHaveAttribute('href', defaultInStockProps.viewAllLink);
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });
  
  it('renders with custom props correctly', () => {
    const customProps = {
      heading: 'Custom InStock Heading',
      subheading: 'Custom InStock Subheading',
      columns: [
        { key: 'name', header: 'Custom Name', align: 'left' }
      ],
      items: [
        {
          id: 'custom-1',
          name: 'Custom Item',
          partNumber: 'CUSTOM-123',
          compatibility: 'Custom Equipment',
          inStock: 'available' as const
        }
      ],
      orderButtonText: 'Custom Order',
      viewAllLinkText: 'Custom View All',
      viewAllLink: '/custom-catalog'
    };
    
    const { container } = render(<InStock {...customProps} />);
    
    // Check if custom heading and subheading are rendered
    expect(screen.getByText(customProps.heading)).toBeInTheDocument();
    expect(screen.getByText(customProps.subheading)).toBeInTheDocument();
    
    // Check if the InStockTable has correct props
    expect(screen.getByText(`Columns: ${customProps.columns.length}`)).toBeInTheDocument();
    expect(screen.getByText(`Items: ${customProps.items.length}`)).toBeInTheDocument();
    expect(screen.getByText(`Order Button Text: ${customProps.orderButtonText}`)).toBeInTheDocument();
    
    // Check if custom stock item is rendered
    const stockItem = screen.getByTestId('stock-item-0');
    expect(stockItem).toBeInTheDocument();
    expect(within(stockItem).getByText(`Name: ${customProps.items[0].name}`)).toBeInTheDocument();
    expect(within(stockItem).getByText(`Part Number: ${customProps.items[0].partNumber}`)).toBeInTheDocument();
    expect(within(stockItem).getByText(`Compatibility: ${customProps.items[0].compatibility}`)).toBeInTheDocument();
    expect(within(stockItem).getByText(`In Stock: ${customProps.items[0].inStock}`)).toBeInTheDocument();
    
    // Check if custom "view all" link is rendered
    expect(screen.getByText(customProps.viewAllLinkText)).toBeInTheDocument();
    expect(screen.getByText(customProps.viewAllLinkText).closest('a')).toHaveAttribute('href', customProps.viewAllLink);
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });
});