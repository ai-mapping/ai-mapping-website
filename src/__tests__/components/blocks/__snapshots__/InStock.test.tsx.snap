// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InStock Component renders with custom props correctly 1`] = `
<div>
  <section
    class="bg-surface py-12"
  >
    <div
      class="container mx-auto px-4 md:px-6 lg:px-8"
    >
      <h2
        class="text-2xl md:text-3xl font-semibold text-center mb-3 text-white"
      >
        Custom InStock Heading
      </h2>
      <p
        class="text-neutral text-center mb-10 max-w-2xl mx-auto"
      >
        Custom InStock Subheading
      </p>
      <div
        class="overflow-x-auto"
      >
        <div
          data-testid="in-stock-table"
        >
          <div>
            Columns: 
            1
          </div>
          <div>
            Items: 
            1
          </div>
          <div>
            Order Button Text: 
            Custom Order
          </div>
          <div>
            <div
              data-testid="stock-item-0"
            >
              <div>
                Name: 
                Custom Item
              </div>
              <div>
                Part Number: 
                CUSTOM-123
              </div>
              <div>
                Compatibility: 
                Custom Equipment
              </div>
              <div>
                In Stock: 
                available
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="text-center mt-8"
      >
        <a
          class="text-primary font-medium hover:text-primary-light transition-colors duration-300"
          data-testid="next-link"
          href="/custom-catalog"
        >
          Custom View All
        </a>
      </div>
    </div>
  </section>
</div>
`;

exports[`InStock Component renders with default props correctly 1`] = `
<div>
  <section
    class="bg-surface py-12"
  >
    <div
      class="container mx-auto px-4 md:px-6 lg:px-8"
    >
      <h2
        class="text-2xl md:text-3xl font-semibold text-center mb-3 text-white"
      >
        Components Library
      </h2>
      <p
        class="text-neutral text-center mb-10 max-w-2xl mx-auto"
      >
        Ready-to-use components for your next project
      </p>
      <div
        class="overflow-x-auto"
      >
        <div
          data-testid="in-stock-table"
        >
          <div>
            Columns: 
            5
          </div>
          <div>
            Items: 
            4
          </div>
          <div>
            Order Button Text: 
            Use Component
          </div>
          <div>
            <div
              data-testid="stock-item-0"
            >
              <div>
                Name: 
                AI Form Builder
              </div>
              <div>
                Part Number: 
                v1.2.0
              </div>
              <div>
                Compatibility: 
                React, Next.js, Vue
              </div>
              <div>
                In Stock: 
                available
              </div>
            </div>
            <div
              data-testid="stock-item-1"
            >
              <div>
                Name: 
                Dynamic Hero Section
              </div>
              <div>
                Part Number: 
                v2.0.1
              </div>
              <div>
                Compatibility: 
                React, Next.js, Astro
              </div>
              <div>
                In Stock: 
                available
              </div>
            </div>
            <div
              data-testid="stock-item-2"
            >
              <div>
                Name: 
                Testimonial Carousel
              </div>
              <div>
                Part Number: 
                v1.5.2
              </div>
              <div>
                Compatibility: 
                React, Vue, Svelte
              </div>
              <div>
                In Stock: 
                available
              </div>
            </div>
            <div
              data-testid="stock-item-3"
            >
              <div>
                Name: 
                AI Content Generator
              </div>
              <div>
                Part Number: 
                v0.9.8
              </div>
              <div>
                Compatibility: 
                React, Next.js
              </div>
              <div>
                In Stock: 
                limited
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="text-center mt-8"
      >
        <a
          class="text-primary font-medium hover:text-primary-light transition-colors duration-300"
          data-testid="next-link"
          href="/components"
        >
          View all available components
        </a>
      </div>
    </div>
  </section>
</div>
`;
